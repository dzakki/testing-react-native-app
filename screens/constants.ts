import { Todos } from "../interfaces/todo";

export const TODOS: Todos = [
  {
    id: 1,
    title: 'learn typescript',
    completed: true,
  },
  {
    id: 2,
    title: 'learn javescript',
    completed: true,
  },
  {
    id: 3,
    title: 'learn PHP',
    completed: false,
  },
  {
    id: 4,
    title: 'learn Vuejs',
    completed: false,
  },
];
