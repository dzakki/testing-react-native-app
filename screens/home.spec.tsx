import axios from 'axios';
import { act, fireEvent, render, screen, userEvent, waitFor } from '@testing-library/react-native';
import Home from '../screens/home';
import { TODOS } from './constants';

jest.mock('axios');

// https://stackoverflow.com/a/53204714
const mockedAxios = axios as jest.Mocked<typeof axios>;

const renderApp = () => {
  mockedAxios.get.mockImplementationOnce(() =>
    Promise.resolve({
      data: TODOS,
    })
  );
  return render(<Home />);
};

describe('Home', () => {
  test('should render content correctly', () => {
    renderApp();

    expect(screen.getByText(TODOS[0].title)).toBeVisible();
    expect(screen.getByText('Todo App!')).toBeVisible();
    expect(screen.getByPlaceholderText('write your todo here...')).toBeVisible();
    expect(screen.getByText('Submit')).toBeVisible();
  });

  test('should add new todo', async () => {
    const event = userEvent.setup();
    jest.useFakeTimers();
    renderApp();

    const inputTodo = screen.getByPlaceholderText('write your todo here...');
    const inputValue = 'learn jest';
    await event.type(inputTodo, inputValue);
    expect(inputTodo).toHaveDisplayValue(inputValue);
    await event.press(screen.getByText('Submit'));

    // helper function from @testing-library/react-native. https://callstack.github.io/react-native-testing-library/docs/api/#helper-functions
    waitFor(() => expect(screen.getByText(inputValue)).toBeVisible());
    jest.useRealTimers();
  });

  test('should match snapshot', () => {
    renderApp();
    expect(screen.toJSON()).toMatchSnapshot();
  })
});

