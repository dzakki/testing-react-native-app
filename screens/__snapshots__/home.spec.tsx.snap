// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home should match snapshot 1`] = `
[
  <View
    style={
      {
        "paddingTop": 16,
        "width": 300,
      }
    }
  >
    <Text
      style={
        {
          "marginBottom": 16,
          "textAlign": "center",
        }
      }
    >
      Todo App!
    </Text>
    <View
      style={
        {
          "flexDirection": "column",
          "flexWrap": "wrap",
          "marginBottom": 8,
        }
      }
    >
      <TextInput
        onChangeText={[Function]}
        placeholder="write your todo here..."
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "marginBottom": 8,
            "padding": 8,
          }
        }
        value=""
      />
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "backgroundColor": "#DDDDDD",
            "borderRadius": 8,
            "marginBottom": 8,
            "opacity": 1,
            "padding": 8,
            "width": "100%",
          }
        }
      >
        <Text
          style={
            {
              "color": "black",
              "textAlign": "center",
            }
          }
        >
          Submit
        </Text>
      </View>
    </View>
  </View>,
  <RCTScrollView
    style={
      {
        "paddingVertical": 16,
        "width": 300,
      }
    }
  >
    <View>
      <RCTScrollView
        data={
          [
            {
              "completed": true,
              "id": 1,
              "title": "learn typescript",
            },
            {
              "completed": true,
              "id": 2,
              "title": "learn javescript",
            },
            {
              "completed": false,
              "id": 3,
              "title": "learn PHP",
            },
            {
              "completed": false,
              "id": 4,
              "title": "learn Vuejs",
            },
          ]
        }
        getItem={[Function]}
        getItemCount={[Function]}
        keyExtractor={[Function]}
        onContentSizeChange={[Function]}
        onLayout={[Function]}
        onMomentumScrollBegin={[Function]}
        onMomentumScrollEnd={[Function]}
        onScroll={[Function]}
        onScrollBeginDrag={[Function]}
        onScrollEndDrag={[Function]}
        removeClippedSubviews={false}
        renderItem={[Function]}
        scrollEventThrottle={0.0001}
        stickyHeaderIndices={[]}
        viewabilityConfigCallbackPairs={[]}
      >
        <View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "backgroundColor": "#eee",
                  "borderRadius": 8,
                  "marginBottom": 8,
                  "padding": 8,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "black",
                  }
                }
              >
                learn typescript
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "backgroundColor": "#f9f9f9",
                  "borderRadius": 8,
                  "marginBottom": 8,
                  "padding": 8,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "black",
                  }
                }
              >
                learn javescript
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "backgroundColor": "#eee",
                  "borderRadius": 8,
                  "marginBottom": 8,
                  "padding": 8,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "black",
                  }
                }
              >
                learn PHP
              </Text>
            </View>
          </View>
          <View
            onFocusCapture={[Function]}
            onLayout={[Function]}
            style={null}
          >
            <View
              style={
                {
                  "backgroundColor": "#f9f9f9",
                  "borderRadius": 8,
                  "marginBottom": 8,
                  "padding": 8,
                  "width": "100%",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "black",
                  }
                }
              >
                learn Vuejs
              </Text>
            </View>
          </View>
        </View>
      </RCTScrollView>
    </View>
  </RCTScrollView>,
]
`;
